File: Lab-main/WPP/ass-7-8/q-2.py
class Queue:
    def __init__(self):
        self.queue = []  # Use a list to represent the queue

    def is_empty(self):
        return len(self.queue) == 0

    def enqueue(self, item):
        self.queue.append(item)  # Add to the rear of the queue
        print(f"Enqueued: {item}")

    def dequeue(self):
        if self.is_empty():
            print("Queue is empty. Cannot dequeue.")
            return None
        else:
            item = self.queue.pop(0)  # Remove from the front
            print(f"Dequeued: {item}")
            return item

    def peek(self): #returns the front element without removing it
        if self.is_empty():
            print("Queue is empty.")
            return None
        else:
            return self.queue[0]

    def size(self):
        return len(self.queue)

    def display(self):
        if self.is_empty():
            print("Queue is empty.")
        else:
            print("Queue:", self.queue)


def display_menu():
    print("\nQueue Operations:")
    print("1. Enqueue")
    print("2. Dequeue")
    print("3. Peek")
    print("4. Size")
    print("5. Display")
    print("6. Exit")


def get_int_input(prompt):
    while True:
        try:
            return int(input(prompt))
        except ValueError:
            print("Invalid input. Please enter an integer.")


def get_data_input(prompt):
    return input(prompt)


queue = Queue()

while True:
    display_menu()
    choice = get_int_input("Enter your choice: ")

    if choice == 1:
        item = get_data_input("Enter item to enqueue: ")
        queue.enqueue(item)
    elif choice == 2:
        queue.dequeue()
    elif choice == 3:
        front_element = queue.peek()
        if front_element is not None:
            print(f"Front element: {front_element}")
    elif choice == 4:
        print("Queue size:", queue.size())
    elif choice == 5:
        queue.display()
    elif choice == 6:
        print("Exiting...")
        break
    else:
        print("Invalid choice. Please try again.")

File: Lab-main/DS/ass-8/q-2.c
#include <stdio.h>
#include <stdlib.h>

#define MAX_SIZE 5

typedef struct {
    int data[MAX_SIZE];
    int front;
    int rear;
} CircularQueue;

CircularQueue* createQueue() {
    CircularQueue* q = (CircularQueue*)malloc(sizeof(CircularQueue));
    q->front = -1;
    q->rear = -1;
    return q;
}

int isFull(CircularQueue* q) {
    return (q->rear + 1) % MAX_SIZE == q->front;
}

int isEmpty(CircularQueue* q) {
    return q->front == -1;
}

void enqueue(CircularQueue* q, int value) {
    if (isFull(q)) {
        printf("Queue is full! Cannot enqueue %d\n", value);
        return;
    }
    if (isEmpty(q)) {
        q->front = 0;
    }
    q->rear = (q->rear + 1) % MAX_SIZE;
    q->data[q->rear] = value;
    printf("Enqueued: %d\n", value);
}

int dequeue(CircularQueue* q) {
    if (isEmpty(q)) {
        printf("Queue is empty!\n");
        return -1;
    }
    int value = q->data[q->front];
    if (q->front == q->rear) { // Last element removed
        q->front = -1;
        q->rear = -1;
    } else {
        q->front = (q->front + 1) % MAX_SIZE;
    }
    printf("Dequeued: %d\n", value);
    return value;
}

int peek(CircularQueue* q) {
    if (isEmpty(q)) {
        printf("Queue is empty!\n");
        return -1;
    }
    return q->data[q->front];
}

int main() {
    CircularQueue* q = createQueue();
    
    enqueue(q, 10);
    enqueue(q, 20);
    enqueue(q, 30);
    enqueue(q, 40);
    enqueue(q, 50); // Should be full after this

    printf("Front element: %d\n", peek(q));
    
    dequeue(q);
    dequeue(q);
    
    enqueue(q, 60);
    enqueue(q, 70);

    printf("Front element after operations: %d\n", peek(q));

    free(q);
    return 0;
}
